src/app/clusters/actions-server/actions-server.h

```diff
@@ -0,0 +1,370 @@
+                        const DataModel::List<const EndpointId> & aEndpointList)
+    {
+        Set(aEpListId, aEpListName, aEpListType, aEndpointList);
+    }
+
+    EndpointListStorage(const EndpointListStorage & aEpList) { *this = aEpList; }
+
+    EndpointListStorage & operator=(const EndpointListStorage & aEpList)
+    {
+        Set(aEpList.endpointListID, aEpList.name, aEpList.type, aEpList.endpoints);
+        return *this;
+    }
+
+    void Set(uint16_t aEpListId, const CharSpan & aEpListName, EndpointListTypeEnum aEpListType,
+             const DataModel::List<const EndpointId> & aEndpointList)
+    {
+        endpointListID    = aEpListId;
+        type              = aEpListType;
+        size_t epListSize = std::min(aEndpointList.size(), ArraySize(mEpList));
+
+        for (size_t index = 0; index < epListSize; index++)
+        {
+            mEpList[index] = aEndpointList[index];
+        }
+        endpoints = DataModel::List<const EndpointId>(Span(mEpList, epListSize));
+        MutableCharSpan mEpListName(mBuffer);
+        CopyCharSpanToMutableCharSpanWithTruncation(aEpListName, mEpListName);
+        name = mEpListName;
+    }
+
+private:
+    char mBuffer[kEndpointListNameMaxSize];
+    EndpointId mEpList[kEndpointListMaxSize];
+};
+
+class ActionsServer : public AttributeAccessInterface, public CommandHandlerInterface
+{
+public:
+    // Register for the Actions cluster on all endpoints.
+    ActionsServer() :
+        AttributeAccessInterface(Optional<EndpointId>::Missing(), Actions::Id),
+        CommandHandlerInterface(Optional<EndpointId>::Missing(), Actions::Id)
+    {}
+    static ActionsServer & Instance();
+
+    /**
+     * @brief
+     *   Called when the state of action is changed.
+     */
+    void OnStateChanged(EndpointId aEndpoint, uint16_t aActionId, uint32_t aInvokeId, ActionStateEnum aActionState);
+
+    /**
+     * @brief
+     *   Called when the action is failed..
+     */
+    void OnActionFailed(EndpointId aEndpoint, uint16_t aActionId, uint32_t aInvokeId, ActionStateEnum aActionState,
+                        ActionErrorEnum aActionError);
+
+    void SetDefaultDelegate(EndpointId aEndpointId, Delegate * aDelegate);
+
+    CHIP_ERROR Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder) override;
+
+    /**
+     * Update an existing action in the action list for the given endpoint.
+     * If the action with the given actionID doesn't exist, returns CHIP_ERROR_NOT_FOUND.
+     *
+     * @param aEndpoint The endpoint ID where the action should be updated
+     * @param aAction The action structure containing the updated action details
+     * @return CHIP_ERROR_INCORRECT_STATE if delegate is null
+     *         CHIP_ERROR_NOT_FOUND if action doesn't exist
+     *         CHIP_NO_ERROR if successful
+     */
+    CHIP_ERROR ModifyActionList(EndpointId aEndpoint, const ActionStructStorage & aAction);
```

Comment:
So this needs to be propagated back to the delegate to actually update the list?  Or is this a notification from the delegate that the list was updated?  I don't understand this API, especially because its implementation does not make sense.

Reply:
OK, then this should be called something like "ActionListModified", if it's a notification, no?  Because right now it sure reads like the cluster implementation is expected to modify the list, which is very much not what you're saying this is supposed to do.

And it it's a notification, why does it need the aAction?  What's it supposed to do with it?

Reply:
Changed an API to a notification, it doesn’t need an `aAction`, removed it.